uhb/mqtt-gdp-gateway.1
.Dd October 4, 2015
.Dt MQTT-GDP-GATEWAY 1
.Os Swarm-GDP
.Sh NAME
.Nm mqtt-gdp-gateway
.Nd copy an MQTT subscription to a log
.Sh SYNOPSIS
.Nm
.Op Fl d
.Op Fl D Ar debug-spec
.Op Fl G Ar router-ip
.Op Fl h Ar interval
.Op Fl F Ar key-file
.Op Fl M Ar broker-addr
.Op Fl s
(
.Ar topic-name
.Ar log-name
) ...
.Sh DESCRIPTION
For each pair of topic and log names,
.Nm
subscribes to the indicated MQTT
.Ar topic-name
and logs all output to the named GDP
.Ar log-name .
The logs must already exist.
Each MQTT message becomes one record in the appropriate log.
.Pp
The
.Ar topic-name
is actually a pattern using standard MQTT syntax:
specifically
.Li +
matches a single lavel of hierarchy and
.Li #
matches all remaining levels of hierarchy
(it must be used as the last item in the topic pattern).
All messages published to a topic that matches the topic pattern
will be logged to the given
.Ar log-name .
.Pp
If the log has a public key included in the metadata
(see
.Xr gcl-create 8 )
then
.Nm
will try to find a corresponding secret key in the filesystem.
It searches for a file named with the internal (256 bit) name of the log
as represented in
base64url
format with a
.Sy \&.pem
file extension containing the secret key
and uses that for signing writes.
See the description of the
.Sy swarm.gdp.crypto.key.path
parameter in
.Sx ADMINISTRATIVE PARAMETERS
for details on the search path.
This search can be overridden using the
.Fl K
flag.
.Pp
Each record logged is formatted as JSON with object tags
.Li topic ,
.Li qos ,
.Li len ,
and
.Li payload .
.Pp
As a special case, if the
.Ar log-name
is a single hyphen, the records are sent to stdout rather than a log.
This is intended for debugging.
.Sh OPTIONS
.Bl -tag
.It Fl d
If set, duplicate records are not logged more than once per minute
(or a value set by the
.Fl h
parameter).
When comparing records, ISO 8601 dates are considered to always match.
.It Fl D Ar debug-spec
Turn on debugging flags using the libep-style format.
Only useful with code in hand.
.It Fl G Ar router-ip
Contact the given IP address (which may be a DNS name)
to get a connection to the GDP routing layer.
This will be replaced by automatic discovery
(for example, using Zeroconf)
soon.
The address may have an optional port number after a colon.
If this is not specified,
the names listed in the
.Va swarm.gdp.routers
runtime administrative parameter
are searched.
.It Fl h Ar interval
A
.Dq heartbeat
interval
forcing records to be output even if duplicates are deleted using the
.Fl d
flag.
Records are sent to the log every
.Ar interval
seconds, even if they are duplicates.
Defaults to the value of the
.Sy swarm.mqtt-gdp-gateway.heartbeat.time
administrative parameter (defaults to 60 seconds).
If zero, duplicates are always deleted,
regardless of how long it has been since the last record was output.
.It Fl K Ar key-file
Names the file containing the secret key to be used for signing.
This must be a
.Sy \&.pem
format file.
If the
.Ar key-file
is a directory,
this parameter overrides the default search path,
so the key file used is based on the name of the log.
.It Fl M Ar broker-addr
Specifies the address of the MQTT broker from which to collect the data.
.It Fl s
If set, the output does not include the wrapping metadata
(topic, qos, and len);
just the content of the payload is output.
.El
.Sh EXIT STATUS
.Bl -tag
.It 0
Success
.It 1
Some failure occurred.
.Sh ADMINISTRATIVE PARAMETERS
.Bl -tag
.It Sy swarm.gdp.crypto.key.path
Controls the search path for secret keys.
Defaults to:
.Bd -unfilled -offset indent -compact
\&.
KEYS
~/.swarm/gdp/keys
/usr/local/etc/swarm/gdp/keys
/etc/swarm/gdp/keys
.Ed
.It Sy swarm.mqtt-gdp-gateway.heartbeat.time
The number of seconds between outputs if duplicates are being deleted using
.Fl d .
Defaults to 60 (seconds).
If zero, no heartbeats are output
(i.e., all duplicate records are deleted,
regardless of how long since the last record was output).
The
.Fl h
flag overrides this parameter.
.El
.\".Sh ENVIRONMENT
.\".Sh FILES
.Sh SEE ALSO
.Xr gcl-create 8
.\".Sh EXAMPLES
.Sh BUGS
Currently,
.Nm
assumes that the payload of each MQTT message
is validly formatted JSON.
This is unrealistic.
.Pp
There should be a way of specifying a separate key file for each GDP log.
.Pp
.Nm
should probably read a configuration file
to get the {topic, log, key} information.


debian/changelog
gdp (0.2-1) UNRELEASED; urgency=low

  * Make multiread startup reliable (was sometimes missing early
    records due to a race condition).
  * Don't hang on a multiread on an empty GCL (it was improperly
    converted to a subscription).
  * Fix a lock ordering problem.

 -- Eric Allman <eric@cs.berkeley.edu>  Wed, 25 Feb 2015 14:15:59 -0800

gdp (0.1-1) UNRELEASED; urgency=low

  * Initial release

 -- Eric Allman <eric@cs.berkeley.edu>  Wed, 11 Feb 2015 11:27:30 -0800



uhb/mqtt-gdp-gateway.c
/* vim: set ai sw=4 sts=4 ts=4 :*/

#include <gdp/gdp.h>
#include <ep/ep_app.h>
#include <ep/ep_dbg.h>
#include <ep/ep_log.h>
#include <ep/ep_string.h>

#include <mosquitto.h>

#include <errno.h>
#include <getopt.h>
#include <sysexits.h>
#include <sys/stat.h>


static EP_DBG	Dbg = EP_DBG_INIT("mqtt-gdp-gateway", "MTQQ to GDP gateway");


bool	SkipMetadata = false;	// don't include topic, qos, and len on output


/*
**  Called by Mosquitto to do logging
*/

void
log_cb(struct mosquitto *mosq,
				void *udata,
				int level,
				const char *str)
{
	const char *level_str = NULL;

	switch (level)
	{
	case MOSQ_LOG_DEBUG:
		//level_str = "debug";
		break;

	case MOSQ_LOG_INFO:
		level_str = "info";
		break;

	case MOSQ_LOG_WARNING:
		level_str = "warning";
		break;

	case MOSQ_LOG_ERR:
		level_str = "error";
		break;

	case MOSQ_LOG_NOTICE:
		level_str = "notice";
		break;

	default:
		fprintf(stderr, "unknown log level %d\n", level);
		level_str = "???";
		break;
	}
	if (level_str != NULL)
		fprintf(stderr, "MOSQ(%s): %s\n", level_str, str);
}


/*
**  GET_TOPIC_INFO --- get information about a given topic
**
**		This implementation is pretty bad.
*/

struct topic_info
{
	const char	*topic_pat;			// pattern for topic
	gdp_gcl_t	*gcl;				// associated GCL
};

struct topic_info	**Topics;		// information about topics
int					NTopics = 0;	// number of topics allocated

struct topic_info *
get_topic_info(const struct mosquitto_message *msg)
{
	int tno;

	ep_dbg_cprintf(Dbg, 2, "get_topic_info(%s): ", msg->topic);
	for (tno = 0; tno < NTopics; tno++)
	{
		struct topic_info *t = Topics[tno];
		if (t == NULL)
			continue;
		bool res;
		int istat = mosquitto_topic_matches_sub(t->topic_pat, msg->topic, &res);
		if (istat != 0 || !res)
			continue;

		// topic matches
		ep_dbg_cprintf(Dbg, 2, "%p (%s)\n", t, t->topic_pat);
		return t;
	}

	// no match
	ep_dbg_cprintf(Dbg, 2, "not found\n");
	return NULL;
}


void
add_topic_info(const char *topic_pat, gdp_gcl_t *gcl)
{
	struct topic_info *t;

	t = ep_mem_zalloc(sizeof *t);
	t->topic_pat = topic_pat;
	t->gcl = gcl;

	Topics = ep_mem_realloc(Topics, (NTopics + 1) * sizeof t);
	Topics[NTopics++] = t;
	ep_dbg_cprintf(Dbg, 3, "add_topic_info(%s): %p\n", topic_pat, t);
}


int
subscribe_to_all_topics(struct mosquitto *mosq, int qos)
{
	int tno;

	for (tno = 0; tno < NTopics; tno++)
	{
		int istat;
		int mid;				// message id --- do we need this?
		struct topic_info *t = Topics[tno];

		if (t == NULL)
			continue;
		ep_dbg_cprintf(Dbg, 5, "subscribe_to_all_topics(%s)\n", t->topic_pat);
		istat = mosquitto_subscribe(mosq, &mid, t->topic_pat, qos);
		if (istat != 0)
			return istat;
	}
	return 0;
}


/*
**  Called by Mosquitto when a message comes in.
*/

void
message_cb(struct mosquitto *mosq,
				void *udata,
				const struct mosquitto_message *msg)
{
	struct topic_info *tinfo;
	const char *payload;

	ep_dbg_cprintf(Dbg, 5, "message_cb: %s @%d => %s\n",
			msg->topic, msg->qos, msg->payload);

	if (msg->payloadlen <= 0)
		payload = "null";
	else
		payload = msg->payload;

	tinfo = get_topic_info(msg);
	if (tinfo == NULL)
	{
		// no place to log this message
		return;
	}

	if (tinfo->gcl != NULL)
	{
		EP_STAT estat;
		gdp_datum_t *datum = gdp_datum_new();

		if (SkipMetadata)
		{
			gdp_buf_printf(gdp_datum_getbuf(datum),
					"%s\n", payload);
		}
		else
		{
			gdp_buf_printf(gdp_datum_getbuf(datum),
					"{topic:\"%s\", qos:%d, len:%d, payload:%s}\n",
					msg->topic, msg->qos, msg->payloadlen, payload);
		}
		estat = gdp_gcl_append(tinfo->gcl, datum);
		if (!EP_STAT_ISOK(estat))
		{
			ep_log(estat, "cannot log MQTT message");
		}
		gdp_datum_free(datum);
	}
	else
	{
		if (SkipMetadata)
		{
			printf("%s\n", payload);
		}
		else
		{
			printf("{topic:\"%s\", qos:%d, len:%d, payload:%s}\n",
					msg->topic, msg->qos, msg->payloadlen, payload);
		}
	}
}


/*
**  Run Mosquitto against a designated broker.
*/

EP_STAT
mosquitto_run(const char *mqtt_broker,
			int subscr_qos,
			void *udata)
{
	struct mosquitto *mosq;
	const char *phase;
	int istat;

	mosquitto_lib_init();

	// get a mosquitto context
	mosq = mosquitto_new(NULL, true, udata);
	if (mosq == NULL)
		goto fail1;

	// can't just pass in NULL for udata, since mosquitto_new automatically
	// maps NULL to mosq
	mosquitto_user_data_set(mosq, udata);

	// set up callbacks
	mosquitto_log_callback_set(mosq, &log_cb);
	mosquitto_message_callback_set(mosq, &message_cb);

	// connect to the broker
	{
		phase = "connect";

		// parse the broker information
		int mqtt_port = 1883;
		char *broker = ep_mem_strdup(mqtt_broker);
		char *p = strrchr(broker, ':');
		if (p != NULL && strchr(p, ']') == NULL)
		{
			*p++ = '\0';
			mqtt_port = atol(p);
		}
		istat = mosquitto_connect(mosq, broker, mqtt_port, 60);
		ep_mem_free(broker);
		if (istat != 0)
			goto fail2;
	}

	phase = "subscribe";
	istat = subscribe_to_all_topics(mosq, subscr_qos);
	if (istat != 0)
		goto fail2;

	phase = "loop";
	istat = mosquitto_loop_forever(mosq, -1, 1);

	// should never get here

fail2:
	if (istat == MOSQ_ERR_ERRNO)
		fprintf(stderr, "mosquitto error in %s: %s\n",
				phase, strerror(errno));
	else
		fprintf(stderr, "mosquitto error in %s: %s\n",
				phase, mosquitto_strerror(istat));
	(void) mosquitto_disconnect(mosq);	// ignore possible MOSQ_ERR_NO_CONN
										// "not connected" error
	mosquitto_destroy(mosq);
	return EP_STAT_ABORT;

fail1:
	fprintf(stderr, "cannot create mosquitto context: %s\n",
			strerror(errno));
	return EP_STAT_ABORT;
}


gdp_gcl_t *
open_signed_log(const char *gdp_log_name, const char *signing_key_file)
{
	EP_STAT estat;
	gdp_name_t gname;
	gdp_gcl_open_info_t *info = NULL;
	gdp_gcl_t *gcl = NULL;

	// make sure we can parse the log name
	estat = gdp_parse_name(gdp_log_name, gname);
	if (!EP_STAT_ISOK(estat))
	{
		char ebuf[60];

		ep_app_error("cannot parse log name %s%s%s: %s",
				EpChar->lquote, gdp_log_name, EpChar->rquote,
				ep_stat_tostr(estat, ebuf, sizeof ebuf));
		goto fail0;
	}

	if (signing_key_file != NULL)
	{
		FILE *fp;
		EP_CRYPTO_KEY *skey;
		struct stat st;

		// set up any open information
		info = gdp_gcl_open_info_new();

		if (stat(signing_key_file, &st) == 0 &&
				(st.st_mode & S_IFMT) == S_IFDIR)
		{
			size_t sz;
			char *p;
			gdp_pname_t pname;

			// we need the printable name for getting the key file name
			gdp_printable_name(gname, pname);

			// find the file in the directory
			sz = strlen(signing_key_file) + sizeof pname + 6;
			p = alloca(sz);
			snprintf(p, sz, "%s/%s.pem", signing_key_file, pname);
			signing_key_file = p;
		}

		fp = fopen(signing_key_file, "r");
		if (fp == NULL)
		{
			ep_app_error("cannot open signing key file %s", signing_key_file);
			goto fail1;
		}

		skey = ep_crypto_key_read_fp(fp, signing_key_file,
				EP_CRYPTO_KEYFORM_PEM, EP_CRYPTO_F_SECRET);
		if (skey == NULL)
		{
			ep_app_error("cannot read signing key file %s", signing_key_file);
			goto fail1;
		}

		estat = gdp_gcl_open_info_set_signing_key(info, skey);
		if (!EP_STAT_ISOK(estat))
		{
fail1:
			if (fp != NULL)
				fclose(fp);
			goto fail0;
		}
	}

	// open a GCL with the provided name
	estat = gdp_gcl_open(gname, GDP_MODE_AO, info, &gcl);
	if (!EP_STAT_ISOK(estat))
	{
		char ebuf[60];

		ep_app_error("cannot open log %s%s%s: %s",
				EpChar->lquote, gdp_log_name, EpChar->rquote,
				ep_stat_tostr(estat, ebuf, sizeof ebuf));
		goto fail0;
	}

fail0:
	if (info != NULL)
		gdp_gcl_open_info_free(info);
	return gcl;
}


void
usage(void)
{
	fprintf(stderr,
			"Usage: %s [-D dbgspec] [-G router_addr] [-K signing_key_file]\n"
			"\t[-M broker_addr] [-q qos] [-s] mqtt_topic gdp_log\n"
			"    -D  set debugging flags\n"
			"    -G  IP host to contact for gdp_router\n"
			"    -K  key file to use to sign GDP log writes\n"
			"    -M  IP host to contact for MQTT broker\n"
			"    -q  MQTT Quality of Service to request\n"
			"    -s  skip metadata in output (topic, qos, len)\n"
			"",
			ep_app_getprogname());
	exit(EX_USAGE);
}


int
main(int argc, char **argv)
{
	EP_STAT estat;
	bool show_usage = false;
	const char *mqtt_broker = NULL;
	int mqtt_qos = 2;
	char *gdp_addr = NULL;
	char *signing_key_file = NULL;
	int opt;
	char ebuf[60];

	while ((opt = getopt(argc, argv, "D:G:K:M:q:s")) > 0)
	{
		switch (opt)
		{
		case 'D':
			ep_dbg_set(optarg);
			break;

		case 'G':
			gdp_addr = optarg;
			break;

		case 'K':
			signing_key_file = optarg;
			break;

		case 'M':
			mqtt_broker = optarg;
			break;

		case 'q':
			mqtt_qos = atoi(optarg);
			break;

		case 's':
			SkipMetadata = true;
			break;

		default:
			show_usage = true;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (show_usage)
		usage();
	if (argc < 2)
	{
		fprintf(stderr, "missing required argument (argc = %d)\n", argc);
		usage();
	}

	// initialize GDP library
	estat = gdp_init(gdp_addr);
	if (!EP_STAT_ISOK(estat))
	{
		ep_app_error("GDP Initialization failed: %s",
				ep_stat_tostr(estat, ebuf, sizeof ebuf));
		exit(EX_UNAVAILABLE);
	}

	// allow thread to settle to avoid interspersed debug output
	ep_time_nanosleep(INT64_C(100000000));

	// open any logs we may use
	for (; argc > 1; argv += 2, argc -= 2)
	{
		struct topic_info *t = ep_mem_zalloc(sizeof *t);
		char *mqtt_topic = argv[0];
		char *gdp_log_name = argv[1];
		gdp_gcl_t *gcl = NULL;

		if (strcmp(gdp_log_name, "-") != 0)
		{
			gcl = open_signed_log(gdp_log_name, signing_key_file);
			if (gcl == NULL)
			{
				// message already given
				show_usage = true;
				continue;
			}
		}

		add_topic_info(mqtt_topic, gcl);
	}

	if (show_usage)
		usage();

	if (mqtt_broker == NULL)
	{
		mqtt_broker = ep_adm_getstrparam("swarm.mqtt-gdp-gateway.broker",
							"127.0.0.1");
	}

	mosquitto_run(mqtt_broker, mqtt_qos, NULL);

	ep_log(estat, "mosquitto_run returned");
	exit(EX_SOFTWARE);
}
